### PR TITLE
Multi arch docker support: add arm/v7 and arm64 to our docker images

### PR DESCRIPTION
**Pull request by maquis196**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -31,3 +43,37 @@ workflows:
...
+      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+      - run: docker context create builder
+      - run: docker buildx create builder --use
+
+  docker_build:
+    description: Builds and pushed images to dockerhub using buildx
+    parameters:
+      dockerfile:
+        type: string
+        default: docker/Dockerfile
```
Comment (richvdh) [inline .circleci/config.yml:68]: it feels like having this (and label) parameterised just adds obfuscation: suggest inlining them.

File: .circleci/config.yml [hunk 2]
```diff
@@ -31,3 +43,37 @@ workflows:
...
+      buildx_version:
+        type: string
+        default: "v0.4.1"
+    steps:
+      - run: apk add --no-cache curl
+      - run: mkdir -vp ~/.docker/cli-plugins/ ~/dockercache
+      - run: curl --silent -L "https://github.com/docker/buildx/releases/download/<< parameters.buildx_version >>/buildx-<< parameters.buildx_version >>.linux-amd64" > ~/.docker/cli-plugins/docker-buildx
+      - run: chmod a+x ~/.docker/cli-plugins/docker-buildx
+      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+      - run: docker context create builder
```
Comment (richvdh) [inline .circleci/config.yml:60]: can you explain why this is required? it appears to create a context that is then unused?

File: .circleci/config.yml [hunk 3]
```diff
@@ -31,3 +43,37 @@ workflows:
...
+    parameters:
+      buildx_version:
+        type: string
+        default: "v0.4.1"
+    steps:
+      - run: apk add --no-cache curl
+      - run: mkdir -vp ~/.docker/cli-plugins/ ~/dockercache
+      - run: curl --silent -L "https://github.com/docker/buildx/releases/download/<< parameters.buildx_version >>/buildx-<< parameters.buildx_version >>.linux-amd64" > ~/.docker/cli-plugins/docker-buildx
+      - run: chmod a+x ~/.docker/cli-plugins/docker-buildx
+      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```
Comment (richvdh) [inline .circleci/config.yml:59]:
```suggestion
      # install qemu links in /proc/sys/fs/binfmt_misc on the docker instance running the circleci job
      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```

File: docker/Dockerfile [hunk 4]
```diff
@@ -20,10 +20,18 @@ FROM docker.io/python:${PYTHON_VERSION}-slim as builder
...
-
 RUN apt-get update && apt-get install -y \
     build-essential \
+    libffi-dev \
+    libjpeg-dev \
     libpq-dev \
+    libssl-dev \
+    libxml++2.6-dev \
+    libxslt1-dev \
+    python3-cryptography \
```
[the remark this answers is not in the document]
Comment (brinsche) [reply docker/Dockerfile:31]: fwiw I just built the 1.19.0 Dockerfile on my Raspberry Pi 4 and only needed to add `libffi-dev` and `libssl-dev`

**Comment by richvdh** [comment]
@maquis196 sorry this now has a bunch of conflicts. could you merge develop in again?

**Comment by richvdh** [comment]
sorry, now your branch has been broken by something else. Could you merge again to pick up https://github.com/matrix-org/synapse/pull/8181 ?

**Comment by maquis196** [comment]
rebased off develop.

File: .circleci/config.yml [hunk 5]
```diff
@@ -29,3 +43,34 @@ workflows:
...
+
+  docker_build:
+    description: Builds and pushed images to dockerhub using buildx
+    parameters:
+      platforms:
+        type: string
+        default: linux/amd64
+      label:
+        type: string
+        default: --label gitsha1=${CIRCLE_SHA1}
```
Comment (richvdh) [inline .circleci/config.yml:72]: I think this is the third time I've asked this: what is the advantage of parameterising this, given it never changes?

File: .circleci/config.yml (hunk 2 above)
Comment (richvdh) [reply .circleci/config.yml:61]: ... but it's not used in the line below?

File: .circleci/config.yml (hunk 5 above)
Comment (maquis196) [reply .circleci/config.yml:72]: youre correct, it was left over when more was parameterised, ill put it back to how it was. Sorry, missed the (and label) bit.

File: .circleci/config.yml (hunk 2 above)
Comment (maquis196) [reply .circleci/config.yml:61]: - run: docker context create builder
      - run: docker buildx create builder --use

We create the context builder, and use the context builder below it, i tried without and it failed, i can do more testing to figure out why that is, appears to be a buildx issue.

Comment (richvdh) [reply .circleci/config.yml:61]: oh sorry, I've just realised that the argument to `docker buildx create` is the name of the context to use.

I think you could probably get away with 

```
run: docker buildx create --use
```

to use the default context. On the other hand it's fine how it is.

Comment (maquis196) [reply .circleci/config.yml:61]: im happy to change it, removes a line of code at least. Lets try it

File: .circleci/config.yml [hunk 6]
```diff
@@ -29,3 +43,31 @@ workflows:
...
+        default: "v0.4.1"
+    steps:
+      - run: apk add --no-cache curl
+      - run: mkdir -vp ~/.docker/cli-plugins/ ~/dockercache
+      - run: curl --silent -L "https://github.com/docker/buildx/releases/download/<< parameters.buildx_version >>/buildx-<< parameters.buildx_version >>.linux-amd64" > ~/.docker/cli-plugins/docker-buildx
+      - run: chmod a+x ~/.docker/cli-plugins/docker-buildx
+      # install qemu links in /proc/sys/fs/binfmt_misc on the docker instance running the circleci job
+      - run: docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
+      - run: docker context create builder
+      - run: docker buildx create builder --use
```
Comment (richvdh) [inline .circleci/config.yml:62]:
```suggestion
      # create a context named `builder` for the builds
      - run: docker context create builder
      # create a buildx builder using the new context, and set it as the default
      - run: docker buildx create builder --use
```

File: .circleci/config.yml (hunk 2 above)
Comment (richvdh) [reply .circleci/config.yml]: ok that looks good.

**Review by richvdh — APPROVED** [review]
lgtm

**Comment by richvdh** [comment]
well that seems to have failed:

```
docker buildx create --use

could not create a builder instance with TLS data loaded from environment. Please use `docker context create <context-name>` to create a context for current environment and then create a builder instance with `docker buildx create <context-name>`
```

could you put the context stuff back in :)

**Comment by maquis196** [comment]
:) good to know for sure that it fails and wasnt just me! So ive reverted mine and added your suggestion for notes.

**Comment by mattcen** [comment]
Hey, folks, I've just stumbled upon this PR while trying to set up a multi-arch build myself (which I've done using GitHub CI/CD: https://github.com/mattcen/synapse/actions/runs/264218899/workflow).
I am particularly looking for a build for linux/arm/v7 to run on Raspberry Pi. Would anybody object to adding `linux/arm/v7` to the `platform:` lines of this PR?

**Comment by maquis196** [comment]
Sure, will certainly need a rebase as well, ill get this in within the next few hours :)

**Comment by maquis196** [comment]
also @richvdh, might be worth adding in to the release notes about the lag between a new tag on github and easily 60-90mins for the non-amd64 images. Especially if as expected, we end up adding ppc and the couple of other archs that dockerhub supports.

**Comment by richvdh** [comment]
It's a bit frustrating that this has been rebased, so I can't easily see what's changed since I last looked at it :/.

**Comment by maquis196** [comment]
rebasing shouldn't matter should it? the last commit - https://github.com/matrix-org/synapse/pull/7921/commits/a09a0beb8a276d5396ecdca6e4104ac5c05bcbde. Shows i simply added v7 to the list and updated the commit log to reflect that.

**Review by richvdh — COMMENTED** [review]
looks good to me. I've kicked off a build at circleci to see what happens (https://app.circleci.com/pipelines/github/matrix-org/synapse/16752/workflows/90fad226-0506-4726-ad16-0c41e640a82a/jobs/25592, for those that have permission to see it).

**Comment by maquis196** [comment]
As a Brit, I'd say, go put the kettle on, this is going to take some time :)

**Comment by maquis196** [comment]
So I saw the build fail, and looking into where it failed it suggests a conflict with arm64 and arm64/v8 images. Its like we can't for whatever reason, use both at the same time. So I've decided to remove the v8 tag and leave it with just arm64 to see if that improves things.

**Comment by mattcen** [comment]
Yeah, that sounds sensible; I was under the impression that arm64 and arm64/v8 were synonymous.

**Comment by richvdh** [comment]
@maquis196 force-pushing makes it hard for people to see what has changed. Please could you reset it to b81ef3e and then merge any changes you want to make?

**Comment by maquis196** [comment]
Sorry, i tried just make a small commit, but being out of sync with the tests meant they failed, so had to rebase and force push. the only change from last time you saw this was - https://github.com/matrix-org/synapse/pull/7921/commits/ab3da5bfe873e6dfce038b8d5fa7168803591183.

**Comment by maquis196** [comment]
and done! not to self, just merge in develop branch in the future.

**Comment by richvdh** [comment]
> and done! not to self, just merge in develop branch in the future.

thanks. It helps me make sure that no unexpected changes are being slipped in with the rebased commits.

**Review by richvdh — APPROVED** [review]
It seems to work \o/. Thanks @maquis196 for all your work on this.

This has landed too late for the 1.21 release cycle, but if anybody wants to test it out before the 1.22 release, `matrixdotorg/synapse:3af9bb7198c64c6e6001b96cbfadc0bc2123d318-multiarch` is a multiarch image of Synapse 1.21.0.

**Comment by richvdh** [comment]
ohh wait. @maquis196: you don't seem to have signed this off. Please can you confirm that you're happy to contribute this work by adding a "Signed-off-by" comment below, as per https://github.com/matrix-org/synapse/blob/develop/CONTRIBUTING.md#sign-off?

**Comment by maquis196** [comment]
how embarrassing, i thought that was in my commit :S (must have been ammended and lost sorry). Also:

Signed-off-by: Christopher May-Townsend (chris@maytownsend.co.uk)

**Pull request merged**